CONTRIBUTING_CODE.md

```diff
@@ -0,0 +1,284 @@
+text, dividing it into a sequence of words and punctuation.
+
+For example:
+
+```ts
+function hello() {
+  console.log("Hi");
+}
+```
+
+Produces a scanner result of:
+`FunctionKeyword, WhitespaceTrivia, Identifier, OpenParenToken, CloseParenToken, WhitespaceTrivia, OpenBraceToken, NewLineTrivia, WhitespaceTrivia, Identifier, DotToken, Identifier, OpenParenToken, StringLiteral, CloseParenToken, SemicolonToken, NewLineTrivia, CloseBraceToken, EndOfFileToken`.
+You can use [this playground plugin](https://www.typescriptlang.org/play?install-plugin=playground-ts-scanner) to
+see the tokens from a scanner for any TS/JS code.
+
+The parser will take that set of syntax tokens and form a syntax tree which is a series of nested nodes starting
+at a `SourceFile`:
+
+```
+SourceFile:
+  - statements: [
+    Function Declaration
+      - name: Identifier
+      - body: Block
+        statements: [
+          ExpressionStatement
+            expression: CallExpress
+              ...
+        ]
+  ]
+```
+
+This is the TypeScript syntax tree, and it is one of the core data models in the compiler as it represents the
+structure of the code in memory. You can explore the TypeScript syntax tree in
+[the Playground](https://www.typescriptlang.org/play/#code/GYVwdgxgLglg9mABACwKYBt1wBQEpEDeAUIohAgM5zqoB0WA5tgEQASMzuA3EQL5A)
+(by turning on the "AST" setting) or in
+[TypeScript AST Viewer](https://ts-ast-viewer.com/#code/GYVwdgxgLglg9mABACwKYBt1wBQEpEDeAUIohAgM5zqoB0WA5tgEQASMzuA3EQL5A).
+
+Common changes to the parser are ones that add TypeScript syntax that does not exist in JavaScript, or adds new JavaScript syntax
+that was approved by the TC39 committee.
+
+#### Type checking
+
+TypeScript's type system works by creating symbols when "something" is declared in a scope. That "something" could
+be a variable, type, interface, function or more. A symbol has a set of flags that indicate what kind(s) of 
+declaration(s) it's for. Symbols are then compared to each other during assignment and are used throughout the type system for
+checking.
+
+For example, this code:
+
+```ts
+function hello() {
+  console.log("Hi");
+}
+
+hello;
+```
+
+creates one new symbol `hello` whose declaration is the function `hello`. The following flag is set:
+`SymbolFlags.Function`. The compiler looks up this symbol whenever the name `hello` appears, like on the last line of the code. 
+You can use [this playground plugin](https://www.typescriptlang.org/play?install-plugin=playground-ts-symbols) to see the
+symbols generated for any TS/JS code.
+
+Creating symbols is the responsibility of the binder, which is the first part of the type checking process.
+
+Type checking in TypeScript happens in the file `checker.ts`, this is a 40k line function which does a huge amount
+of work. The checker works primarily by recursively diving into, known as "walking", the syntax tree, and making assertions about the
```

Comment:
The type checker walking the syntax tree should be mentioned at the beginning of the section since the reader has just learned what the syntax tree is.